### ISSUE TITLE
[Error][Swift][deliver][submit] Use of Advertising Identifier (IDFA) is required to submit

### ISSUE THREAD

**Issue by rogerluan** [fastlane/fastlane#16799]
### New Release Candidate Issue Checklist

- [x] Updated fastlane to the latest release candidate version (2.150.0.rcX)
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Information

- Breaking version: 2.151.1
- Last working version: 🤷‍♀️ 
- State of app: `Prepare for Submission`
- First version of app?: No

### Issue Description

I've always specified the `submissionInformation` in my `Deliverfile.swift` file and it used to work (pre-2.150.0). I haven't been able to submit a build for review (from 0 to hero with no human interaction) ever since 2.150.0, and this is the latest error I've been facing:

`Deliverfile.swift`:
```swift
    /// Extra information for the submission (e.g. compliance specifications, IDFA settings).
    var submissionInformation: Any? {
        return [
            "add_id_info_limits_tracking" : true,
            "add_id_info_serves_ads" : false,
            "add_id_info_tracks_action" : true,
            "add_id_info_tracks_install" : true,
            "add_id_info_uses_idfa" : true,
            "content_rights_has_rights" : true,
            "content_rights_contains_third_party_content" : true,
            "export_compliance_platform" : "ios",
            "export_compliance_compliance_required" : false,
            "export_compliance_encryption_updated" : false,
            "export_compliance_app_type" : nil,
            "export_compliance_uses_encryption" : false,
            "export_compliance_is_exempt" : false,
            "export_compliance_contains_third_party_cryptography" : false,
            "export_compliance_contains_proprietary_cryptography" : false,
            "export_compliance_available_on_french_store" : false
        ]
    }
```

This PR might be related to this: https://github.com/fastlane/fastlane/pull/16706

##### Complete output when running fastlane, including the stack trace and command used

```

+--------------------------------------+-------------------------------------------+
|                             deliver 2.151.1 Summary                              |
+--------------------------------------+-------------------------------------------+
| username                             | apple.ci@zilly.email                      |
| app_identifier                       | com.zilly.zilly                           |
| app_version                          | 2.30.0                                    |
| platform                             | ios                                       |
| edit_live                            | false                                     |
| use_live_version                     | false                                     |
| metadata_path                        | ../tellus-ios-app-store-metadata/metadata |
| skip_binary_upload                   | true                                      |
| skip_screenshots                     | true                                      |
| skip_metadata                        | false                                     |
| skip_app_version_update              | false                                     |
| force                                | true                                      |
| overwrite_screenshots                | false                                     |
| submit_for_review                    | true                                      |
| reject_if_possible                   | false                                     |
| automatic_release                    | false                                     |
| phased_release                       | false                                     |
| reset_ratings                        | false                                     |
| run_precheck_before_submit           | true                                      |
| precheck_default_rule_level          | warn                                      |
| individual_metadata_items            | []                                        |
| ignore_language_directory_validation | false                                     |
| precheck_include_in_app_purchases    | true                                      |
| screenshots_path                     | ./fastlane/screenshots                    |
+--------------------------------------+-------------------------------------------+

[…trimmed…]
…
…
[20:28:57]: fastlane finished with errors

[!] Use of Advertising Identifier (IDFA) is required to submit
Add information to the :submission_information option...
  Docs: http://docs.fastlane.tools/actions/deliver/#compliance-and-idfa-settings
  Example: submission_information: { add_id_info_uses_idfa: false }
  Example: submission_information: {
    add_id_info_uses_idfa: true,
    add_id_info_limits_tracking: false,
    add_id_info_serves_ads: false,
    add_id_info_uses_idfa: false,
    add_id_info_tracks_install: false
  }
  Example CLI:
    --submission_information "{\"add_id_info_uses_idfa\": false}"
```

### Environment

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                           |
| --------------------------- | --------------------------------------------------------------- |
| OS                          | 10.15.5                                                         |
| Ruby                        | 2.6.5                                                           |
| Bundler?                    | true                                                            |
| Git                         | git version 2.24.3 (Apple Git-128)                              |
| Installation Source         | ~/Documents/Projects/tellus-ios/.vendor/ruby/2.6.0/bin/fastlane |
| Host                        | Mac OS X 10.15.5 (19F101)                                       |
| Ruby Lib Dir                | ~/.rbenv/versions/2.6.5/lib                                     |
| OpenSSL Version             | OpenSSL 1.1.1d  10 Sep 2019                                     |
| Is contained                | false                                                           |
| Is homebrew                 | false                                                           |
| Is installed via Fabric.app | false                                                           |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                     |
| Xcode Version               | 11.5                                                            |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE |             |   |

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.151.1 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| bundler                   | 2.1.4        |
| rake                      | 13.0.1       |
| CFPropertyList            | 3.0.2        |
| ZenTest                   | 4.12.0       |
| RubyInline                | 3.12.5       |
| concurrent-ruby           | 1.1.6        |
| i18n                      | 0.9.5        |
| minitest                  | 5.14.1       |
| thread_safe               | 0.3.6        |
| tzinfo                    | 1.2.7        |
| activesupport             | 4.2.11.3     |
| public_suffix             | 4.0.5        |
| addressable               | 2.7.0        |
| httpclient                | 2.8.3        |
| json                      | 2.3.1        |
| algoliasearch             | 1.27.3       |
| atomos                    | 0.1.3        |
| aws-eventstream           | 1.1.0        |
| aws-partitions            | 1.339.0      |
| aws-sigv4                 | 1.2.1        |
| jmespath                  | 1.4.0        |
| aws-sdk-core              | 3.103.0      |
| aws-sdk-kms               | 1.36.0       |
| aws-sdk-s3                | 1.73.0       |
| babosa                    | 1.0.3        |
| claide                    | 1.0.3        |
| fuzzy_match               | 2.0.4        |
| nap                       | 1.1.0        |
| netrc                     | 0.11.0       |
| ffi                       | 1.13.1       |
| ethon                     | 0.12.0       |
| typhoeus                  | 1.4.0        |
| cocoapods-core            | 1.9.1        |
| cocoapods-deintegrate     | 1.0.4        |
| cocoapods-downloader      | 1.3.0        |
| cocoapods-plugins         | 1.0.0        |
| cocoapods-search          | 1.0.0        |
| cocoapods-stats           | 1.1.0        |
| cocoapods-trunk           | 1.5.0        |
| cocoapods-try             | 1.2.0        |
| colored2                  | 3.1.2        |
| escape                    | 0.0.4        |
| fourflusher               | 2.3.1        |
| gh_inspector              | 1.1.3        |
| molinillo                 | 0.6.6        |
| ruby-macho                | 1.4.0        |
| nanaimo                   | 0.2.6        |
| xcodeproj                 | 1.17.0       |
| cocoapods                 | 1.9.1        |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| cocoapods-binary          | 0.4.4        |
| dotenv                    | 2.7.5        |
| osx_keychain              | 1.0.2        |
| cocoapods-keys            | 2.2.1        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| declarative               | 0.0.20       |
| declarative-option        | 0.1.0        |
| digest-crc                | 0.6.1        |
| unf_ext                   | 0.0.7.7      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20190701 |
| emoji_regex               | 3.0.0        |
| excon                     | 0.75.0       |
| multipart-post            | 2.0.0        |
| faraday                   | 1.0.1        |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 1.0.0        |
| fastimage                 | 2.1.7        |
| jwt                       | 2.2.1        |
| memoist                   | 0.16.2       |
| multi_json                | 1.14.1       |
| os                        | 1.1.0        |
| signet                    | 0.14.0       |
| googleauth                | 0.13.0       |
| mini_mime                 | 1.0.2        |
| uber                      | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| google-api-client         | 0.38.0       |
| google-cloud-env          | 1.3.2        |
| google-cloud-errors       | 1.0.1        |
| google-cloud-core         | 1.5.0        |
| google-cloud-storage      | 1.26.2       |
| mini_magick               | 4.10.1       |
| plist                     | 3.5.0        |
| rubyzip                   | 2.3.0        |
| security                  | 0.1.3        |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.8        |
| slack-notifier            | 2.3.2        |
| terminal-notifier         | 2.0.0        |
| unicode-display_width     | 1.7.0        |
| terminal-table            | 1.8.0        |
| tty-screen                | 0.8.0        |
| tty-cursor                | 0.7.1        |
| tty-spinner               | 0.9.3        |
| word_wrap                 | 1.0.0        |
| xcpretty-travis-formatter | 1.0.0        |
| mustache                  | 1.1.1        |
| open4                     | 1.3.4        |
| redcarpet                 | 3.5.0        |
| sassc                     | 2.4.0        |
| sqlite3                   | 1.4.2        |
| liferaft                  | 0.0.6        |
| xcinvoke                  | 0.3.0        |
| jazzy                     | 0.13.4       |
| psych                     | 3.0.3        |
</details>

*generated on:* **2020-07-08**
</details>

### Mentioning

cc: @joshdholtz

**Comment by joshdholtz** [comment]
@rogerluan Thanks for the report! It seems like `submissionInformation` is getting sent over from the Swift interface to the Ruby runner 🤔 I will try to take a look at this tomorrow to see what’s going on!

**Comment by rogerluan** [comment]
Thanks for the prompt response again! 🚀

**Comment by rogerluan** [comment]
In this issue I noticed that there was an issue with the example copy printed out ☝️ So I opened a fix here https://github.com/fastlane/fastlane/pull/16807

**Comment by rogerluan** [comment]
@joshdholtz I took a stab at fixing this issue today, and I'm happy to say I resolved the issue! Although I'm not sure if a fix is needed...

So due to lack of better debugging knowledge 🤷‍♀️ I added dozens of UI.message logs around the ruby code, printed the stack trace until I tracked down where the `submission_information` was being lost... Apparently it wasn't being received even in the lowest level of the socket layer, in `socket_server.rb`:

```ruby
def parse_and_execute_command(command_string: nil)
  UI.message("Roger's logs parse_and_execute_command(command_string: #{command_string}") // It wasn't present here
  command = CommandParser.parse(json: command_string)
  UI.message("Roger's logs command: #{command.to_yaml}")
  case command
  when ControlCommand
    return handle_control_command(command)
  when ActionCommand
    return handle_action_command(command)
  end

  # catch all
  raise "Command #{command} not supported"
end
```

So I started thinking that was too weird, and decided to check the Swift code more closely... Until I realized this sneaky change:

https://github.com/fastlane/fastlane/commit/fb4a23c5b19607a8eefa81ad5a83099b573ed425#diff-62f1882363af9a2f50f7e0603fe971ddR518

And I realized that I was using the old type notation (`var submissionInformation: Any?`) as you can see in this issue's original description.

# tl;dr: This change in my Deliverfile.swift fixed the issue:

```diff
    /// Extra information for the submission (e.g. compliance specifications, IDFA settings).
-    var submissionInformation: Any? {
+    var submissionInformation: [String: Any]? {
        return [
            "add_id_info_limits_tracking" : true,
            "add_id_info_serves_ads" : false,
            "add_id_info_tracks_action" : true,
            "add_id_info_tracks_install" : true,
            "add_id_info_uses_idfa" : true,
            "content_rights_has_rights" : true,
            "content_rights_contains_third_party_content" : true,
            "export_compliance_platform" : "ios",
            "export_compliance_compliance_required" : false,
            "export_compliance_encryption_updated" : false,
-            "export_compliance_app_type" : nil,
+            "export_compliance_app_type" : "",
            "export_compliance_uses_encryption" : false,
            "export_compliance_is_exempt" : false,
            "export_compliance_contains_third_party_cryptography" : false,
            "export_compliance_contains_proprietary_cryptography" : false,
            "export_compliance_available_on_french_store" : false
        ]
    }
```

(note that I don't know what to do with `export_compliance_app_type`, nor know what it is for, but using `""` worked 🙈 )

Doing that made the `submission_information.[…]`-related entries show up in `deliver` Summary table again, and fastlane finished with success 🚀 

Now, not sure how you wanna address this issue... Maybe leaving as is isn't too bad, people will definitely face issues but they can find this issue in GitHub. Other 2 ways to fix it would be make `Any` readable again, or revert the code to receive `Any` instead of the Swift dictionary... But sounds like these alternatives are too much hassle for a non-issue 🤷

Hope this helps you @joshdholtz , since my issue was resolved, feel free to close this issue if no change is going to happen 👍 
Cheers

**Comment by joshdholtz** [comment]
@rogerluan Oops! I missed this response 😱 We should probably change `Any` to `Any?` but maybe @minuscorp might have a better thought? 😇

**Comment by rogerluan** [comment]
@joshdholtz no worries! I think that omitting the key that would have a `nil` value would work (I haven't tested doing that in my Swift config), and essentially it's the same behavior as `[String:Any?]`, but requires no changes in the fastlane 😇 What do you think?

**Comment by rogerluan** [comment]
I can test this on my next release btw, just to make sure

**Comment by rogerluan** [comment]
Or I could actually use the right `export_compliance_app_type` value, the problem is that idk what to put there, cuz idk the possible values (there's no documentation for this) 🙈 
I think this falls into the "split `submission_information` into separate fields" task hehe

**Comment by minuscorp** [comment]
If you want to try to put `nil` change the type to [String: Any?] directly, idk if that works when being parsed to Ruby though

**Comment by joshdholtz** [comment]
@minuscorp What is the different between `Any?` and `AnyObject?`? 😇 

@rogerluan I completely agree with the split submission information part. I do really want to focus on a rework here 🙃 Maybe we close this since things will get reworked in #16854?

**Comment by minuscorp** [comment]
Gotta be `Any`, `AnyObject` is a metatype protocol which all classes conform to, so Int, Double, String, are structs cannot be `AnyObject` my bad, my brain is dead today.

**Comment by rogerluan** [comment]
Sure, I'll test that case too @minuscorp !
@joshdholtz I'd be in favour of that, maybe we'd be wasting time here. I just wanna test those cases to make sure we understand what's the current state of things 👍

**Comment by rogerluan** [comment]
Closing this issue.
Current workaround: https://github.com/fastlane/fastlane/issues/16799#issuecomment-657170824
Future improvement: https://github.com/fastlane/fastlane/issues/16854
